File: Dokie/ViewControllers/Folder/FolderViewController.m
//
//  FolderViewController.m
//  Dokie
//
//  Created by Ashamaz Shidov on 28/12/2019.
//  Copyright © 2019 Ashamaz Shidov. All rights reserved.
//

#import "FolderViewController.h"
#import "FolderViewModel.h"
#import "Logger.h"
#import "FileNameTableViewCell.h"
#import "FolderNameTableViewCell.h"
#import "UI.h"

@interface FolderViewController () <UITableViewDelegate, UITableViewDataSource>

@property (nonatomic, strong) FolderViewModel *viewModel;

@end

@implementation FolderViewController

- (instancetype)initWithViewModel:(FolderViewModel *)viewModel {
    ParameterAssert(viewModel);
    
    self = [super init];
    if (self) {
        self.viewModel = viewModel;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setup];
}

- (void)setup {
    @weakify(self);
    
    [[RACObserve(self, viewModel.folderName) deliverOnMainThread] subscribeNext:^(NSString *folderName) {
        @strongify(self);
        self.navigationItem.title = folderName;
    }];
    
    UIBarButtonItem *addButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:nil action:nil];
    self.navigationItem.rightBarButtonItems = @[addButton];
    addButton.rac_command = [[RACCommand alloc] initWithSignalBlock:^RACSignal * _Nonnull(id _) {
        @strongify(self);
        [self.viewModel addButtonPressed];
        return [RACSignal empty];
    }];

    if (self.viewModel.showMenuNavButton) {
        UIBarButtonItem *menuButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"settings_small.png"] style:UIBarButtonItemStylePlain target:nil action:nil];
        menuButton.rac_command = [[RACCommand alloc] initWithSignalBlock:^RACSignal * _Nonnull(id _) {
            @strongify(self);
            [self.viewModel menuPressed];
            return [RACSignal empty];
        }];
        self.navigationItem.leftBarButtonItems = @[menuButton];
    }
    UITableView *tableView = [UI tableView];
    tableView.delegate = self;
    tableView.dataSource = self;
    [self.view addSubview:tableView];
    
    UILabel *noContentLabel = [UI label];
    noContentLabel.text = NSLocalizedString(@"Folder is empty", @"Folder is empty content text");
    noContentLabel.alpha = 0;
    noContentLabel.textAlignment = NSTextAlignmentCenter;
    [self.view addSubview:noContentLabel];
    
    NSDictionary *views = NSDictionaryOfVariableBindings(tableView, noContentLabel);
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|[tableView]|" options:0 metrics:nil views:views]];
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|[tableView]|" options:0 metrics:nil views:views]];
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|[noContentLabel]|" options:0 metrics:nil views:views]];
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|[noContentLabel]|" options:0 metrics:nil views:views]];
    
    UILongPressGestureRecognizer *longPress = [[UILongPressGestureRecognizer alloc] initWithTarget:tableView action:nil];
    longPress.minimumPressDuration = 1.0;
    @weakify(tableView);
    [longPress.rac_gestureSignal subscribeNext:^(__kindof UIGestureRecognizer *gestureRecognizer) {
        @strongify(self, tableView);

        if (gestureRecognizer.state == UIGestureRecognizerStateBegan) {
            CGPoint p = [gestureRecognizer locationInView:tableView];
            NSIndexPath *indexPath = [tableView indexPathForRowAtPoint:p];
            if (indexPath.section == 0) {
                [self.viewModel longPressedSubfolderAtIndexPath:indexPath];
            }
            else {
                [self.viewModel longPressedDocumentAtIndexPath:indexPath];
            }
        }
    }];
    [tableView addGestureRecognizer:longPress];
    
    @weakify(noContentLabel);
    [[[RACSignal combineLatest:@[RACObserve(self, viewModel.subfolders),
                                 RACObserve(self, viewModel.documents)]]
      deliverOnMainThread]
     subscribeNext:^(RACTuple *tuple) {
        @strongify(tableView, noContentLabel);
        RACTupleUnpack(NSArray *subfolders, NSArray *documents) = tuple;
        noContentLabel.alpha = subfolders.count + documents.count > 0 ? 0 : 1;
        [tableView reloadData];
    }];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 2;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return section == 0 ? self.viewModel.subfolders.count : self.viewModel.documents.count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 50;
}

static NSString *const subfolderCellIdentifier = @"subfolderCellIdentifier";
static NSString *const documentCellIdentifier = @"documentCellIdentifier";

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    @weakify(self);
    if (indexPath.section == 0) {
        FolderNameTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:subfolderCellIdentifier];
        if (!cell) {
            cell = [[FolderNameTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:subfolderCellIdentifier];
            [cell setSelectionStyle:UITableViewCellSelectionStyleNone];
        }
        cell.editCommand = [[RACCommand alloc] initWithSignalBlock:^RACSignal * _Nonnull(id _) {
            @strongify(self);
            [self.viewModel longPressedSubfolderAtIndexPath:indexPath];
            return [RACSignal empty];
        }];
        cell.name = self.viewModel.subfolders[indexPath.row];
        return cell;
    }
    else {
        FileNameTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:documentCellIdentifier];
        if (!cell) {
            cell = [[FileNameTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:documentCellIdentifier];
            [cell setSelectionStyle:UITableViewCellSelectionStyleNone];
        }
        cell.name = self.viewModel.documents[indexPath.row];
        cell.editCommand = [[RACCommand alloc] initWithSignalBlock:^RACSignal * _Nonnull(id _) {
            @strongify(self);
            [self.viewModel longPressedDocumentAtIndexPath:indexPath];
            return [RACSignal empty];
        }];
        return cell;
    }
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    if (indexPath.section == 0) {
        [self.viewModel didSelectSubfolderAtIndexPath:indexPath];
    }
    else {
        [self.viewModel didSelectDocumentAtIndexPath:indexPath];
    }
}

@end

File: Dokie/ViewControllers/Folder/FolderViewController.swift
//
//  FolderViewController.swift
//  Dokie
//
//  Created by Ashamaz Shidov on 25.05.2020.
//  Copyright © 2020 Ashamaz Shidov. All rights reserved.
//

import UIKit
import ReactiveCocoa
import ReactiveSwift

class FolderViewController: BaseViewController, UISearchDisplayDelegate {
    
    private let viewModel: FolderViewModel!
    private let noContentLabel = UI.label()
    private let searchBar = UISearchBar()
    
    init(viewModel: FolderViewModel) {
        self.viewModel = viewModel
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        setup()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        searchBar.resignFirstResponder()
    }
    
    private func setup() {
        viewModel
            .folderName
            .producer
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithValues { [unowned self] name in
                self.navigationItem.title = name
        }
        
        let addButton = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(addButtonPressed))
        viewModel
            .showAddButtonBar
            .producer
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithValues { [unowned self, addButton] show in
                self.navigationItem.rightBarButtonItems = show ? [addButton] : []
        }

        if viewModel.showMenuNavBarButton {
            let menuButton = UIBarButtonItem(image: UIImage(named: "settings_small.png"), style: .plain, target: self, action: #selector(menuButtonPressed))
            navigationItem.leftBarButtonItems = [menuButton]
        }
        
        searchBar.translatesAutoresizingMaskIntoConstraints = false
        searchBar.barTintColor = Appearance.backgroundColor
        searchBar.searchTextField.backgroundColor = Appearance.tagBackgroundColor
        searchBar.placeholder = String.localized("Search tags")
        searchBar.searchTextField.tintColor = .white
        searchBar.searchTextField.textColor = .white
        searchBar.delegate = self
        view.addSubview(searchBar)
        
        let tableView = UI.tableView()
        tableView.delegate = self
        tableView.dataSource = self
        view.addSubview(tableView)

        noContentLabel.alpha = 0
        noContentLabel.textAlignment = .center
        view.addSubview(noContentLabel)
        viewModel
            .noContentText
            .producer
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithValues { [unowned self] text in
                self.noContentLabel.text = text
        }

        let metrics = ["searchBarHeight"    : 44,
                       "sideMargin"         : 8]
        let views = ["searchBar" : searchBar, "tableView" : tableView, "noContentLabel" : noContentLabel]
        view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|[searchBar(searchBarHeight)][tableView]|", options: NSLayoutConstraint.FormatOptions(rawValue: 0), metrics: metrics, views: views))
        view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|[tableView]|", options: NSLayoutConstraint.FormatOptions(rawValue: 0), metrics: metrics, views: views))
        view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-sideMargin-[searchBar]|", options: NSLayoutConstraint.FormatOptions(rawValue: 0), metrics: metrics, views: views))
        view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|[noContentLabel]|", options: NSLayoutConstraint.FormatOptions(rawValue: 0), metrics: metrics, views: views))
        view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|[noContentLabel]|", options: NSLayoutConstraint.FormatOptions(rawValue: 0), metrics: metrics, views: views))
        
        let longPress = UILongPressGestureRecognizer(target: nil, action: nil)
        longPress.minimumPressDuration = 1
        longPress.reactive.stateChanged.producer.startWithValues { [unowned self, tableView] gesture in
            guard gesture.state == .began else { return }
            let p = gesture.location(in: tableView)
            
            guard let indexPath = tableView.indexPathForRow(at: p) else { return }
            
            if indexPath.section == 0 {
                self.viewModel.longPressedSubfolder(at: indexPath)
            }
            else {
                self.viewModel.longPressedDocument(at: indexPath)
            }
        }
        tableView.addGestureRecognizer(longPress)
        
        viewModel
            .subfolders
            .producer
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithValues { [unowned tableView] _ in
                tableView.reloadData()
        }
        
        viewModel
            .documents
            .producer
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithValues { [unowned tableView] _ in
                tableView.reloadData()
        }
    }
    
    @objc private func addButtonPressed() {
        viewModel.addButtonPressed()
    }
    
    @objc private func menuButtonPressed() {
        viewModel.menuPressed()
    }
    
}

extension FolderViewController : UITableViewDelegate, UITableViewDataSource {
    
    var subfolderCellIdentifier: String { "subfolderCellIdentifier" }
    var documentCellIdentifier: String { "documentCellIdentifier" }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if indexPath.section == 0 {
            let cell: FolderNameTableViewCell
            if let dequeuedCell = tableView.dequeueReusableCell(withIdentifier: subfolderCellIdentifier) as? FolderNameTableViewCell {
                cell = dequeuedCell
            }
            else {
                cell = FolderNameTableViewCell(style: .default, reuseIdentifier: subfolderCellIdentifier)
                cell.selectionStyle = .none
            }
            
            let action = Action<(), UIButton, Never> { _ -> SignalProducer<UIButton, Never> in
                return SignalProducer { [unowned self] (observer, disposable) in
                    self.viewModel.longPressedSubfolder(at: indexPath)
                    observer.sendCompleted()
                }
            }
            
            cell.editComand = CocoaAction(action)
            cell.name.value = viewModel.subfolders.value[indexPath.row]
            
            return cell
        }
        else {
            let cell: FileNameTableViewCell
            if let dequeuedCell = tableView.dequeueReusableCell(withIdentifier: documentCellIdentifier) as? FileNameTableViewCell {
                cell = dequeuedCell
            }
            else {
                cell = FileNameTableViewCell(style: .default, reuseIdentifier: documentCellIdentifier)
                cell.selectionStyle = .none
            }
            
            let action = Action<(), UIButton, Never> { _ -> SignalProducer<UIButton, Never> in
                return SignalProducer { [unowned self] (observer, disposable) in
                    self.viewModel.longPressedDocument(at: indexPath)
                    observer.sendCompleted()
                }
            }
            
            cell.editComand = CocoaAction(action)
            cell.name.value = viewModel.documents.value[indexPath.row]
            
            return cell
        }
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        noContentLabel.alpha = viewModel.documents.value.count + viewModel.subfolders.value.count > 0 ? 0 : 1
        
        return 2
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        section == 0 ? viewModel.subfolders.value.count : viewModel.documents.value.count
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        if indexPath.section == 0 {
            viewModel.didSelectSubfolder(at: indexPath)
        }
        else {
            viewModel.didSelectDocument(at: indexPath)
        }
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        50
    }
}

extension FolderViewController : UISearchBarDelegate {
    
    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
        viewModel.searchText.value = searchText
    }

}

extension FolderViewController : UIScrollViewDelegate {
    
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        searchBar.resignFirstResponder()
    }
    
}
